llvm/lib/Transforms/Vectorize/LoopIdiomVectorize.cpp

```diff
@@ -679,7 +657,12 @@ void LoopIdiomVectorize::transformByteCompare(GetElementPtrInst *GEPA,
   BasicBlock *Header = CurLoop->getHeader();
   BranchInst *PHBranch = cast<BranchInst>(Preheader->getTerminator());
   IRBuilder<> Builder(PHBranch);
+
+  // Safeguard to check if we build the correct DomTree with DTU.
+  auto CheckDTU = llvm::make_scope_exit(
+      [this]() { assert(DT->verify() && "Ill-formed DomTree built by DTU"); });
```

Comment:
Given DTU only flushes when the object is destroyed at the end of the function can we guarantee that `DT->verify()` is called after the flush? Perhaps the answer is to do an explicit `flush()` at the end followed by the assert?

Reply:
DTU is declared after this `make_scope_exit` object, according C++ standard destructors will be invoked in the reverse of their declared order, so I believe this `make_scope_exit` callback is always called after `~DTU()`.

Reply:
OK thanks for explaining. Perhaps it's just me that didn't find this obvious, but I wonder if it would be useful to leave a brief comment explaining how this works? For example,

```
  // Safeguard to check if we build the correct DomTree with DTU. In accordance
  // with C++ rules, destructors are called in reverse order so the verify()
  // occurs after ~DTU.
```